### PR TITLE
Appoptics-papertrail mixer adapter integrated with istio mixer

### PR DESCRIPTION
**Pull request by girishranganathanswi**
This change is for a new mixer adapter for sending metrics to appoptics (https://www.appoptics.com) and logs to papertrail (https://papertrailapp.com).

### CODE REVIEW

**Comment by girishranganathanswi** [comment]
I signed it!

File: mixer/adapter/appoptics/metrics_handler.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	if strings.TrimSpace(cfg.AppopticsAccessToken) != "" {
+		lc := appoptics.NewClient(cfg.AppopticsAccessToken, env.Logger())
+
+		go promadapter.BatchMeasurements(&loopFactor, prepChan, pushChan, stopChan, env.Logger())
+		go promadapter.PersistBatches(&loopFactor, lc, pushChan, stopChan, errorChan, env.Logger())
+		go promadapter.ManagePersistenceErrors(&loopFactor, errorChan, stopChan, env.Logger())
+	} else {
+		go func() {
+			// to drain the channel
+			for range prepChan {
```
Comment (mandarjog) [inline mixer/adapter/appoptics/metrics_handler.go]: If accessToken is not provided preChan is drained? Is there anything in that channel to be drained at this point?

File: mixer/adapter/appoptics/config/config.proto [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    
+    string appoptics_access_token = 1;
+    
+    string papertrail_url = 2;
+
+    string papertrail_local_retention = 3;
+    
+    message MetricInfo {
+        string name = 1;
+	    string instance_name = 2;
```
Comment (mandarjog) [inline mixer/adapter/appoptics/config/config.proto]: looks like the proto file is not formatted correctly.

File: DEV-GUIDE.md [hunk 3]
```diff
@@ -201,8 +201,8 @@ git push origin my-feature
 
 ### Creating a pull request
 
-1. Visit https://github.com/$GITHUB_USER/istio
-2. Click the "Compare & pull request" button next to your "my-feature" branch.
+1. Visit https://github.com/$GITHUB_USER/istio if you created a fork in your own github repostiory, or https://github.com/istio/istio and navigate to your branch (e.g. "my-feature").
```
Comment (mandarjog) [inline DEV-GUIDE.md]: Please send this change as a separate PR.

**Comment by mandarjog** [comment]
@girishranganathanswi  thanks for the PR.
This is a 2.2k line PR excluding the generated proto file, is it possible to break it up into multiple pieces? Are all paper trail, metrics and logs required to be part of the same PR ?

File: mixer/adapter/appoptics/metrics_handler.go (hunk 1 above)
Comment (girishranganathanswi) [reply mixer/adapter/appoptics/metrics_handler.go]: If access token is not provided, this is setup to drain metrics that are getting sent to the prepchan which will happen in the handleMetric method.

**Comment by girishranganathanswi** [comment]
Sorry, yes it is a composite adapter so both app optics and papertrail changes are needed. But I will close this and send a new PR. There are a few things going on.
